### PR TITLE
z1651

### PR DESCRIPTION
**Pull request by kawaiiyoshiyuki**

### CODE REVIEW

File: src/cliAction/serviceLogs/handler_ws.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+func (h *Handler) receiveHandler(connection *websocket.Conn, format, mode string) {
+	defer close(done)
+
+	for {
+		_, msg, err := connection.ReadMessage()
+		if err != nil {
+			// websocket close err (appears on expiration of token)
+			closeErr := strings.Contains(err.Error(), "websocket: close")
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) || closeErr {
```
Comment (vnarek) [inline src/cliAction/serviceLogs/handler_ws.go:87]: Why is there the `closeErr` check? If I understand the `IsUnexpectedCloseError` right every error which is not `websocket.CloseGoingAway` or `websocket.CloseAbnormalClosure` should return `true`. By checking for closeErr these two error codes are then also matching this condition. This condition should be equal to `websocket.IsUnexpectedCloseError(err)` I think

File: src/cliAction/serviceLogs/handler_ws.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			}
+			finishedByUser := strings.Contains(err.Error(), "use of closed network connection")
+			if !finishedByUser {
+				errMsg := fmt.Errorf("%s %s\n", i18n.LogReadingFailed, err.Error())
+				fmt.Println(errMsg)
+			}
+			return
+		}
+
+		if strings.Contains(string(msg), "{\"items\"") && !strings.Contains(string(msg), "{\"items\":[]}") {
```
Comment (vnarek) [inline src/cliAction/serviceLogs/handler_ws.go:99]: I would just try to parse the json and wouldn't check if the structure contains given string. User could have `{ "item": ["{\"items\": []}"]` as a response. :grin: I know it's unlikely but still.

File: src/cliAction/serviceLogs/handler_ws.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			lastMsgId = updateLastMsgId(msg)
```
Comment (vnarek) [inline src/cliAction/serviceLogs/handler_ws.go:100]: It's better to not have global variables in code generally. You lose the ability to track variable changes effectively. Whenever it's possible avoid having them.

File: src/cliAction/serviceLogs/handler_ws.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	interrupt chan os.Signal
+	lastMsgId string
+)
+
+func (h *Handler) getLogStream(
+	ctx context.Context, inputs InputValues, uri, query, containerId, logServiceId, projectId string,
+) error {
+	url := updateUri(uri, query)
+
+	interrupt = make(chan os.Signal, 1)    // Channel to listen for interrupt signal to terminate gracefully
```
Comment (vnarek) [inline src/cliAction/serviceLogs/handler_ws.go:26]: global variable

File: src/cliAction/serviceLogs/handler_printLogs.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+func (h *Handler) printLogs(ctx context.Context, inputs InputValues, containerId, logServiceId, projectId string) error {
+	method, url, err := h.getServiceLogResData(ctx, h.sdkConfig, projectId)
+	if err != nil {
+		return err
+	}
+
+	query := makeQueryParams(inputs, logServiceId, containerId)
+
+	if inputs.mode == RESPONSE {
+		err = getLogs(ctx, method, HTTP+url+query, inputs.format, inputs.formatTemplate, inputs.mode)
```
Comment (vnarek) [inline src/cliAction/serviceLogs/handler_printLogs.go:17]: Shouldn't `https` protocol be used instead?

File: src/cliAction/serviceLogs/handler_ws.go (hunk 1 above)
Comment (kawaiiyoshiyuki) [reply src/cliAction/serviceLogs/handler_ws.go:87]: dunno why but it didn't work like that, that's why I added this check

**Pull request merged**